### PR TITLE
EKS: Support tagging node group's underlying ASG

### LINKED ISSUE #2881: EKS: Support tagging node group's underlying ASG
/kind feature

**Describe the solution you'd like**
There is an [open issue](https://github.com/aws/containers-roadmap/issues/1541) on EKS where EKS doesn’t support tagging the underlying ASGs, however it’s quite important for many organizations for compliance, cost. It would be great to add such support in CAPA, where user can specify an optional `AdditionalAsgTags` field in `AWSManagedMachinePool` and the node group controller will reconcile the underlying ASG tags.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by richardchen331**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
There is an [open issue](https://github.com/aws/containers-roadmap/issues/1541) on EKS where EKS doesn’t support tagging the underlying ASGs, however it’s quite important for many organizations for compliance, cost. This PR adds such support in CAPA, where user can specify an optional `AdditionalAsgTags` field in `AWSManagedMachinePool` and the node group controller will reconcile the underlying ASG tags.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2881

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Support tagging EKS node group's underlying ASG
```

### CODE REVIEW

File: exp/api/v1alpha3/awsmanagedmachinepool_types.go [hunk 1]
```diff
@@ -73,6 +73,10 @@ type AWSManagedMachinePoolSpec struct {
 	// +optional
 	AdditionalTags infrav1alpha3.Tags `json:"additionalTags,omitempty"`
 
+	// AdditionalAsgTags is an optional set of tags to add to the underlying ASG resources managed by the AWS provider
+	// +optional
+	AdditionalAsgTags infrav1alpha3.Tags `json:"additionalAsgTags,omitempty"`
```
Comment (shivi28) [inline exp/api/v1alpha3/awsmanagedmachinepool_types.go:78]: Shall we rename it `AdditionalASGTags`. Same at other places also

File: pkg/cloud/services/eks/nodegroup.go [hunk 2]
```diff
@@ -486,6 +486,10 @@ func (s *NodegroupService) reconcileNodegroup() error {
 		return errors.Wrapf(err, "failed to reconcile nodegroup tags")
 	}
 
+	if err := s.reconcileAsgTags(ng); err != nil {
```
Comment (shivi28) [inline pkg/cloud/services/eks/nodegroup.go:489]: Rename it to `reconcileASGTags`

File: pkg/cloud/services/eks/tags.go [hunk 3]
```diff
@@ -19,9 +19,12 @@ package eks
 import (
 	"fmt"
 
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+
```
Comment (shivi28) [inline pkg/cloud/services/eks/tags.go:23]: remove this space

File: pkg/cloud/services/eks/tags.go [hunk 4]
```diff
@@ -73,6 +76,35 @@ func (s *NodegroupService) reconcileTags(ng *eks.Nodegroup) error {
...
+		return nil
+	}
+	trueVal := true
+	asgType := "auto-scaling-group"
+	for _, asg := range ng.Resources.AutoScalingGroups {
+		input := &autoscaling.CreateOrUpdateTagsInput{}
+		for k, v := range s.scope.ManagedMachinePool.Spec.AdditionalAsgTags {
+			// The k/vCopy is used to address the "Implicit memory aliasing in for loop" issue
+			// https://stackoverflow.com/questions/62446118/implicit-memory-aliasing-in-for-loop
+			kCopy := k
```
Comment (shivi28) [inline pkg/cloud/services/eks/tags.go:90]: Can we index the ranged map as given in the stackoverflow link?
```
for key := range s.scope.ManagedMachinePool.Spec.AdditionalAsgTags {
    key:= key (Can remove it and check if it still throws error)
    input.Tags = append(input.Tags, &autoscaling.Tag{
		Key:               &key,
		PropagateAtLaunch: &trueVal,
		ResourceId:        asg.Name,
		ResourceType:      &asgType,
		Value:             & s.scope.ManagedMachinePool.Spec.AdditionalAsgTags[key],
	})
}
if _, err := s.AutoscalingClient.CreateOrUpdateTags(input); err != nil {
	return errors.Wrap(err, "failed to reconcile AutoScalingGroup tags for nodegroup")
}
```

File: pkg/cloud/services/eks/tags.go [hunk 5]
```diff
@@ -73,6 +76,35 @@ func (s *NodegroupService) reconcileTags(ng *eks.Nodegroup) error {
...
+			vCopy := v
+			input.Tags = append(input.Tags, &autoscaling.Tag{
+				Key:               &kCopy,
+				PropagateAtLaunch: &trueVal,
```
Comment (shivi28) [inline pkg/cloud/services/eks/tags.go:94]: Can use `aws.Bool()` function from library `"github.com/aws/aws-sdk-go/aws"`. It saves one variable `trueVal`

File: pkg/cloud/services/eks/tags.go [hunk 6]
```diff
@@ -73,6 +76,35 @@ func (s *NodegroupService) reconcileTags(ng *eks.Nodegroup) error {
...
+				ResourceId:        asg.Name,
+				ResourceType:      &asgType,
```
Comment (shivi28) [inline pkg/cloud/services/eks/tags.go:96]: similarly you can do for `asgtype` using func `pointer.String("foo")` from lib `"k8s.io/utils/pointer"`

File: exp/api/v1alpha3/awsmanagedmachinepool_types.go (hunk 1 above)
Comment (richardchen331) [reply exp/api/v1alpha3/awsmanagedmachinepool_types.go]: Sure, renamed to `AdditionalASGTags` :)

File: pkg/cloud/services/eks/nodegroup.go (hunk 2 above)
Comment (richardchen331) [reply pkg/cloud/services/eks/nodegroup.go]: Done

File: pkg/cloud/services/eks/tags.go (hunk 3 above)
Comment (richardchen331) [reply pkg/cloud/services/eks/tags.go]: done

File: pkg/cloud/services/eks/tags.go (hunk 5 above)
Comment (richardchen331) [reply pkg/cloud/services/eks/tags.go]: Code updated

File: pkg/cloud/services/eks/tags.go (hunk 6 above)
Comment (richardchen331) [reply pkg/cloud/services/eks/tags.go]: done

**Comment by richardchen331** [comment]
Updated code to remove unspecified ASG tags as well

File: pkg/cloud/services/eks/tags.go (hunk 4 above)
Comment (richardchen331) [reply pkg/cloud/services/eks/tags.go:134]: I tried to change to this but got error `invalid operation: cannot take address of s.scope.ManagedMachinePool.Spec.AdditionalAsgTags[key] (map index expression of type string)`, so I'll keep it as it for now

**Comment by richardchen331** [comment]
Hi @shivi28 @randomvariable could you help take a look at this PR when you get chance? Thank you! :)

File: pkg/cloud/services/eks/nodegroup.go [hunk 7]
```diff
@@ -18,6 +18,7 @@ package eks
 
 import (
 	"fmt"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/awserrors"
```
Comment (shivi28) [inline pkg/cloud/services/eks/nodegroup.go:21]: enter is required on line 21

File: pkg/cloud/services/eks/tags.go [hunk 8]
```diff
@@ -68,11 +71,103 @@ func getTagUpdates(currentTags map[string]string, tags map[string]string) (untag
 	return untagKeys, newTags
 }
 
+func getASGTagUpdates(clusterName string, currentTags map[string]string, tags map[string]string) (tagsToDelete map[string]string, tagsToAdd map[string]string) {
+	officialASGTagsByEKS := []string{
+		"eks:cluster-name",
```
Comment (shivi28) [inline pkg/cloud/services/eks/tags.go:76]: Would you like to store these tags in a separate variable and then directly use them here?

File: pkg/cloud/services/eks/tags.go [hunk 9]
```diff
@@ -68,11 +71,103 @@ func getTagUpdates(currentTags map[string]string, tags map[string]string) (untag
...
+		"eks:nodegroup-name",
```
Comment (shivi28) [inline pkg/cloud/services/eks/tags.go:77]: Same for all tags

**Comment by shivi28** [comment]
Suggested few minor changes @richardchen331.

**Comment by richardchen-db** [comment]
> Suggested few minor changes @richardchen331.

Thanks for your review @shivi28 ! I updated the PR suggested by you, please let me know if I understand your comments correctly / if you have other feedback. Thanks!

**Comment by richardchen331** [comment]
> Suggested few minor changes @richardchen331.

Thanks for your review @shivi28 ! I updated the PR suggested by you, please let me know if I understand your comments correctly / if you have other feedback. Thanks!

**Comment by shivi28** [comment]
@richardchen331 tests are failing. Try `make lint` in your local

**Comment by shivi28** [comment]
Can you please squash all your commits

**Comment by richardchen331** [comment]
> Can you please squash all your commits

Yep, squashed all the commits : )

File: exp/api/v1alpha4/awsmanagedmachinepool_types.go [hunk 10]
```diff
@@ -83,6 +83,10 @@ type AWSManagedMachinePoolSpec struct {
 	// +optional
 	AdditionalTags infrav1alpha4.Tags `json:"additionalTags,omitempty"`
 
+	// AdditionalASGTags is an optional set of tags to add to the underlying ASG (auto scaling group) resources managed by the AWS provider
+	// +optional
+	AdditionalASGTags infrav1alpha4.Tags `json:"additionalASGTags,omitempty"`
```
Comment (richardcase) [inline exp/api/v1alpha4/awsmanagedmachinepool_types.go:88]: Do we need the  new `AdditionalASGTags` field or can we use the existing `AdditionalTags`?

Comment (richardcase) [reply exp/api/v1alpha4/awsmanagedmachinepool_types.go:88]: Is there a scenario where a user (lets say you) where you would want to apply different additional tags to the ASG compared to the EKS control plane?

Comment (richardchen331) [reply exp/api/v1alpha4/awsmanagedmachinepool_types.go:88]: The `AdditionalTags` are applied to the NodeGroup resource, and the `AdditionalASGTags` are intended to be applied to the underlying ASG resource. I'm not sure if we want users to be able to specify them differently (In my case, what I really care about is the tags applied to the ASG for compliance reasons, and they can be the same as the ones applied to the NodeGroup). 

Please let me know what you think. Thanks!

Comment (richardcase) [reply exp/api/v1alpha4/awsmanagedmachinepool_types.go:88]: In other controllers we use `AdditionalTags` and apply these across different AWS resources.  I think for consistency and to stop us setting a precedent for creating more`AdditionalXXXTags` fields in the future that we should change the  reconciliation of ASG tags to use the existing `AdditionalTags` field. wdyt?

Comment (richardchen331) [reply exp/api/v1alpha4/awsmanagedmachinepool_types.go:88]: That sounds good to me! I'll make the chagne.

**Comment by richardchen331** [comment]
@richardcase PR updated to reuse `AdditionalTags`, could you help take a final look? Thanks!

**Comment by richardchen331** [comment]
@richardcase it looks like the e2e test is somehow flaky (timed out after 1800s) looking at the job history (https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-cluster-api-provider-aws-e2e-eks). Should I retry a few times?

**Comment by richardcase** [comment]
Very strange `failed to find releases tagged with a valid semantic version number`. Will rung again to make sure then we can investigate if it fails again.

/test pull-cluster-api-provider-aws-e2e-eks

**Comment by richardcase** [comment]
Same issue. I will investigate

**Comment by richardchen331** [comment]
e2e tests passed after rebasing the PR. @richardcase do you mind taking a final look at this PR? Thanks!

**Comment by richardcase** [comment]
Probable a flake, so:

/test pull-cluster-api-provider-aws-e2e-eks

**Comment by richardcase** [comment]
@richardchen331 - Looking at the CAPA logs for the e2e failure i see:

```
I0117 11:20:31.049787       1 awsmanagedmachinepool_controller.go:175]  "msg"="Reconciling AWSManagedMachinePool"  
I0117 11:20:31.907542       1 tags.go:126]  "msg"="Reconciling ASG tags"  "cluster-name"="cluster-m075u4" "nodegroup-name"="eks-nodes-3eljk4_cluster-m075u4-pool-0"
I0117 11:20:32.057082       1 awsmanagedmachinepool_controller.go:195]  "msg"="Reconciling deletion of AWSManagedMachinePool" 
```

From the logs it looks like its starting reconciling the ASG tags but then:

- fails during reconciling
- Or more likely we are hittinh a timeout that is causing the cluster to be deleted at the time of reconciling the ASG tags

The timeout is probably more likely due to this in the logs:

```
   Timed out after 1800.000s.
      
  Expected
      <int>: 1
  to equal
      <int>: 2 
```

**Comment by richardcase** [comment]
I've increased the timeout by 5 mins to see it's purely unfortunate timing or if the extra time reveals any errors.

**Comment by richardchen331** [comment]
@richardcase  Looking at the e2e test log it looks like that:
- The test timed out "Waiting for the machine pool workload nodes to exist" after "Scaling the machine pool up" (log [here](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/2884/pull-cluster-api-provider-aws-e2e-eks/1483346306503020544/build-log.txt), test implementation [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/2dabf93aedc43216c66772e7d2f27c9656bf20e3/test/e2e/suites/managed/machine_pool.go#L90))
- It timed out because of MachinePool.Status.ReadyReplicas (1) != MachinePool.Spec.Replicas (2)
- Looking at the [MP resource](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/2884/pull-cluster-api-provider-aws-e2e-eks/1483346306503020544/artifacts/clusters/bootstrap/resources/eks-nodes-gngtfw/MachinePool/cluster-cv7161-pool-0.yaml) and the [AWSMMP resource](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/2884/pull-cluster-api-provider-aws-e2e-eks/1483346306503020544/artifacts/clusters/bootstrap/resources/eks-nodes-gngtfw/AWSManagedMachinePool/cluster-cv7161-pool-0.yaml) it looks that the cloud resource indeed scaled up, just that one of the node is not ready. 
- This doesn't seems to be related to the code change. I'm going to rerun the test and see what happens.

**Comment by richardcase** [comment]
This is strange as the periodic e2e for EKS are passing every night.

I can't see why this change would cause the e2e to fail.

I'll run the e2e locally on your branch to see if it highlights anything.

**Comment by richardchen331** [comment]
@richardcase I think I found the culprit... So EKS nodes relies on the "kubernetes.io/cluster/cluster-name:owned" tag in order to join the cluster (https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html#worker-node-fail), however we are incorrectly updating the ASG tags because `ManagedMachinePoolScope.ClusterName()` returns the Cluster CR name instead of the actual EKS cluster name. I updated the PR and let's see if tests can pass.

**Comment by richardcase** [comment]
Ah, we've had that issue in other places. Fingers crossed it wired works and the e2e pass. If it didn't I'll run them locally this morning

**Comment by richardcase** [comment]
Thats good the eks e2e [passed](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/2884/pull-cluster-api-provider-aws-e2e-eks/1484860625388572672)

**Comment by richardcase** [comment]
The tests are passing, i think this is good to go:

/lgtm
/approve

**Pull request merged**